hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioOutboundPipeline.java

```diff
@@ -206,73 +255,118 @@ public void process() throws Exception {
                 break;
             case BLOCKED:
-                // pipeline is blocked; no point in receiving OP_WRITE events.
-                unregisterOp(OP_WRITE);
+                postProcessBlocked();
                 break;
             default:
                 throw new IllegalStateException();
         }
     }
 
-    /**
-     * Tries to unschedule this pipeline.
-     * <p/>
-     * It will only be unscheduled if:
-     * - there are no pending frames.
-     * <p/>
-     * If the outputBuffer is dirty then it will register itself for an OP_WRITE since we are interested in knowing
-     * if there is more space in the socket output buffer.
-     * If the outputBuffer is not dirty, then it will unregister itself from an OP_WRITE since it isn't interested
-     * in space in the socket outputBuffer.
-     * <p/>
-     * This call is only made by the owning IO thread.
-     */
-    private void unschedule() throws IOException {
-        // since everything is written, we are not interested anymore in write-events, so lets unsubscribe
+    private void postProcessBlocked() throws IOException {
+        // pipeline is blocked; no point in receiving OP_WRITE events.
         unregisterOp(OP_WRITE);
-        // So the outputBuffer is empty, so we are going to unschedule the pipeline.
-        scheduled.set(false);
 
-        if (writeQueue.isEmpty() && priorityWriteQueue.isEmpty()) {
-            // there are no remaining frames, so we are done.
-            return;
+        for (; ; ) {
+            State state = scheduled.get();
+            if (state == State.SCHEDULED || state == State.BLOCKED) {
+                if (!scheduled.compareAndSet(state, State.BLOCKED)) {
+                    continue;
+                }
+
+                if (priorityWriteQueue.isEmpty() || !scheduled.compareAndSet(State.BLOCKED, State.SCHEDULED)) {
+                    break;
+                }
+
+                owner().addTaskAndWakeup(this);
+                break;
+            } else if (state == State.RESCHEDULE) {
+                scheduled.set(State.SCHEDULED);
+                owner().addTaskAndWakeup(this);
+                break;
+            } else {
+                throw new IllegalStateException("unexpected state:" + state);
+            }
         }
+    }
 
-        // So there are frames, but we just unscheduled ourselves. If we don't try to reschedule, then these
-        // Frames are at risk not to be send.
-        if (!scheduled.compareAndSet(false, true)) {
-            //someone else managed to schedule this OutboundHandler, so we are done.
-            return;
+    private void postProcessDirty() throws IOException {
+        // pipeline is dirty, so register for an OP_WRITE to write more data.
+        registerOp(OP_WRITE);
+
+        if (writeThroughEnabled && !(Thread.currentThread() instanceof NioThread)) {
+            // there was a write through. Changing the interested set of the selection key
```

Comment:
Can you use `owner().addTaskAndWakeup(this);` here instead?

Reply:
It is purely about the selector. IMHO it is best to explain your intent instead of going through a higher level call and hope to accomplish the same (potentially bypassing the behavior you want to get triggered).